### PR TITLE
feat: sort and filter functionality and labelled notes listing

### PR DESCRIPTION
**Pull request by PuneetP16**
## VIDEO PREVIEW

https://user-images.githubusercontent.com/91944618/161525573-79a57d54-a3a4-4c73-b4cf-6ec11cfaea17.mp4

## NOTE
I have hardcoded note data for review purposes so that all the filter's functionality could be checked easily. If you edit or add a new note all the hardcoded data would be destroyed as it syncs with the backend.

So check move to archive and trash, change color, and edit functionality on newly added notes only.

Files to check:
FilterPanel.jsx
filterContext.js
noteContext.js
Label.jsx
filterOperations.js
getFormattedDate.js

### CODE REVIEW

File: src/components/FilterPanel/FilterPanel.jsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+			payload: priorityEvent.target.value,
+		});
+	};
+
+	const dispatchByLabel = (labelEvent) => {
+		filterDispatch({
+			type: "BY_LABEL",
+			payload: labelEvent.target.value,
+		});
+	};
```
Comment (m26unkwn) [inline src/components/FilterPanel/FilterPanel.jsx:39]: written very well wonderful variable naming.

File: src/components/FilterPanel/FilterPanel.jsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+	return (
+		<form className="filter_panel">
+			{/* sortBy */}
+			<h3 className="h5 filter__head">Filters</h3>
```
Comment (m26unkwn) [inline src/components/FilterPanel/FilterPanel.jsx:44]: Remove this commented code

File: src/components/NoteLabel/NoteLabel.jsx [hunk 3]
```diff
@@ -7,32 +7,36 @@ export const NoteLabel = () => {
...
 
 	const setNoteLabelHandler = (e) => {
 		e.preventDefault();
-		console.log(noteLabel);
-		noteDispatch({
-			type: "SET_LABEL",
-			payload: noteLabel,
-		});
-		setNoteLabel("");
+		if (noteLabel.trim() === "") {
```
Comment (m26unkwn) [inline src/components/NoteLabel/NoteLabel.jsx:17]: what are you doing here.

remove this if it not useful.

image

File: src/components/NoteListing/NoteListing.jsx [hunk 4]
```diff
@@ -54,20 +57,22 @@ export const NoteListing = ({ list, isPinSection, isTrash, isArchive }) => {
...
 
 	return list.map((note, i) => {
-		const { _id, title, body, noteColor, dateCreated, labels } = note;
+		const { _id, title, body, noteColor, dateCreated, labels, priority } = note;
 
 		const toggleArchiveBtn = isArchive ? bxIcons.archiveOut : bxIcons.archiveIn;
 		const toggleTrashBtn = isTrash ? <MdiRestore /> : bxIcons.trashAlt;
 		const togglePinBtn = isPinSection ? bxIcons.pinned : bxIcons.pin;
 
 		return (
```
Comment (m26unkwn) [inline src/components/NoteListing/NoteListing.jsx:66]: You can create a component here.
then you can pass these as props to it.
This is quite confusing.

File: src/contexts/filterContext.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+	const value = {
+		filterDispatch,
+		sortByDate,
+		category,
+		selectedLabel,
+		byPriority,
+		initialFilterState,
+	};
+
```
Comment (m26unkwn) [inline src/contexts/filterContext.js:71]: Good way to pass the values to the provider.

File: src/contexts/magnificentContext.js [hunk 6]
```diff
@@ -17,15 +18,17 @@ export const MagnificentProvider = ({ children }) => {
...
+						<AuthProvider>
+							<UserProvider>
+								<NoteProvider>
+									<ThemeProvider>
+										<LoaderProvider>{children}</LoaderProvider>
+									</ThemeProvider>
+								</NoteProvider>
+							</UserProvider>
+						</AuthProvider>
+					</FilterProvider>
```
Comment (m26unkwn) [inline src/contexts/magnificentContext.js:31]: Amazing.🔥

This is a good way of using component and then using it as global Provider .

Nice code.

File: src/utilities/filterOperations.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { getFormattedDate } from "./getFormattedDate";
+
+export const sortItByDate = (notesList, sortByDate) => {
+	const data = [...notesList];
+
+	const ascending = (a, b) => {
+		const previousTime = getFormattedDate(a.dateCreated).getTime();
+		const nextTime = getFormattedDate(b.dateCreated).getTime();
+		return previousTime - nextTime;
+	};
```
Comment (m26unkwn) [inline src/utilities/filterOperations.js:10]: very well written.

**Review by m26unkwn — APPROVED** [review]
Overall your code is clean and well written.

You can checkout the comments.

File: src/pages/Label/Label.jsx [hunk 8]
```diff
@@ -1,6 +1,43 @@
...
+				{labelsList.length > 1 ? (
+					labelsList.map((label, index) => {
+						const filteredLabelList = noteList.filter((note) =>
+							note.labels.includes(label)
+						);
+						const camelCaseLabel = `${label.split("")[0].toUpperCase()}${label
+							.split("")
+							.splice(1)
+							.join("")}`;
+						return (
```
Comment (m26unkwn) [inline src/pages/Label/Label.jsx:24]: This is good work to write the camelCase name.

You can extract it from here.

File: src/components/FilterPanel/FilterPanel.jsx [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+export const FilterPanel = () => {
+	const {
+		sortByDate,
+		byPriority,
+		filterDispatch,
+		selectedLabel,
+	} = useFilter();
+	const { noteState } = useNote();
+	const { labelsList } = noteState;
```
Comment (manojsarna) [inline src/components/FilterPanel/FilterPanel.jsx:12]: Suggestion : You can do this one line while destructuring!!

File: src/components/FilterPanel/FilterPanel.css [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	display: flex;
+	flex-flow: row nowrap;
+	align-items: baseline;
+	gap: 7px;
+}
+
+.select {
+	cursor: pointer;
+	/* min-width: 16rem; */
+	width: min-content;
```
Comment (dhruvin25799) [inline src/components/FilterPanel/FilterPanel.css:38]: Remove commented code if not needed.

File: src/components/FilterPanel/FilterPanel.css [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+.select:focus {
+	border-color: var(--COLOR-PRIMARY-THEME);
+}
+
+.select--sort_by {
+}
+
+.option {
```
Comment (dhruvin25799) [inline src/components/FilterPanel/FilterPanel.css:57]: empty CSS classes, if you aren't planning on using them, they can be removed.

File: src/pages/Label/Label.jsx (hunk 8 above)
Comment (manojsarna) [inline src/pages/Label/Label.jsx:24]: Suggestion : Create a custom function which will return camelCase String :)

File: src/components/FilterPanel/bck.css [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+	gap: 1rem;
+	width: fit-content;
+	padding: 1.5rem 3rem;
+	border-radius: 8px;
+	transition: all 1s cubic-bezier(0.165, 0.84, 0.44, 1);
+	position: fixed;
+	/* position: absolute; */
+	/* left: -27rem;
+	bottom: -35rem; */
+	left: 8rem;
```
Comment (dhruvin25799) [inline src/components/FilterPanel/bck.css:14]: Remove commented code here

File: src/components/FilterPanel/bck.css [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+.select:focus {
+	border-color: var(--COLOR-PRIMARY-THEME);
+}
+
+.select--sort_by {
+}
+
+.option {
+}
```
Comment (dhruvin25799) [inline src/components/FilterPanel/bck.css:64]: Empty CSS classes can be removed.

File: src/reducers/noteReducer.js [hunk 14]
```diff
@@ -1,50 +1,61 @@
 export const noteReducer = (noteState, action) => {
-	let { noteObj, noteList, trashList, archiveList } = noteState;
+	let { noteObj, noteList, trashList, archiveList, labelsList } = noteState;
 	let { type, payload } = action;
```
Comment (manojsarna) [inline src/reducers/noteReducer.js:3]: Suggestion : naming conventions can be better

File: src/reducers/noteReducer.js [hunk 15]
```diff
@@ -1,50 +1,61 @@
...
 
+	const updateLabelsList = (...lists) => {
+		const allList = Array.from(lists).flat();
```
Comment (manojsarna) [inline src/reducers/noteReducer.js:6]: Suggestion : either use the name  updateLabels or updateLabelList

**Review by dhruvin25799 — APPROVED** [review]
Good clean code.
Good UI, great implementation of the functionality.
What I like: 
Consistent and easy to understand variable naming.
Looks good to merge.

UI Suggestion : 
The placeholders for the Sign-Up Form have full opacity and looks like some input is already there, either change it to a lighter color and reduce the opacity to not confuse the user.

File: src/reducers/noteReducer.js [hunk 16]
```diff
@@ -1,50 +1,61 @@
...
 		case "COLOR":
 			noteObj = { ...noteObj, noteColor: payload };
-			return { noteObj, noteList, trashList, archiveList };
+			return { noteObj, noteList, trashList, archiveList, labelsList };
 
 		case "SAVE":
+			console.log(payload)
 			noteList = [...payload];
-			return { noteObj, noteList, trashList, archiveList };
+			return { noteObj, noteList, trashList, archiveList, labelsList };
```
Comment (manojsarna) [inline src/reducers/noteReducer.js:27]: Suggestion : Update the noteState using {...state, state[noteObj]:updatedValue}

File: src/utilities/filterOperations.js [hunk 17]
```diff
@@ -0,0 +1,34 @@
+import { getFormattedDate } from "./getFormattedDate";
+
+export const sortItByDate = (notesList, sortByDate) => {
+	const data = [...notesList];
+
+	const ascending = (a, b) => {
+		const previousTime = getFormattedDate(a.dateCreated).getTime();
+		const nextTime = getFormattedDate(b.dateCreated).getTime();
```
Comment (manojsarna) [inline src/utilities/filterOperations.js:8]: Appreciation : Good work with the formattedDate Function

File: src/utilities/filterOperations.js [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+	if (byPriority === "All") {
+		return filteredList;
+	}
+	return filteredList.filter((note) => note.priority === byPriority);
+};
+
+export const filterByLabel = (filteredList, selectedLabel) => {
+	if (selectedLabel === "All") {
+		return filteredList;
+	}
```
Comment (manojsarna) [inline src/utilities/filterOperations.js:32]: Appreciation : Good work with filters

File: src/components/NoteListing/NoteListing.jsx [hunk 19]
```diff
@@ -54,20 +57,22 @@ export const NoteListing = ({ list, isPinSection, isTrash, isArchive }) => {
 	};
 
 	return list.map((note, i) => {
-		const { _id, title, body, noteColor, dateCreated, labels } = note;
+		const { _id, title, body, noteColor, dateCreated, labels, priority } = note;
 
 		const toggleArchiveBtn = isArchive ? bxIcons.archiveOut : bxIcons.archiveIn;
 		const toggleTrashBtn = isTrash ? <MdiRestore /> : bxIcons.trashAlt;
 		const togglePinBtn = isPinSection ? bxIcons.pinned : bxIcons.pin;
```
Comment (manojsarna) [inline src/components/NoteListing/NoteListing.jsx:65]: Suggestion : Convert this into a component

image

**Review by manojsarna — APPROVED** [review]
UI Review : 
1. Button cursor pointer is overflowing to the whole row. Do Check !!
If I click in the row of the button , onClick triggers on the whole row!! Check This!!

image

2. Give CTA priority to Login first then Signup
image

**Pull request merged**